homeassistant/components/control4/media_player.py

```diff
@@ -0,0 +1,406 @@
+
+from homeassistant.components.media_player import (
+    MediaPlayerDeviceClass,
+    MediaPlayerEntity,
+    MediaPlayerEntityFeature,
+    MediaPlayerState,
+    MediaType,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from . import Control4Entity
+from .const import CONF_DIRECTOR, CONF_DIRECTOR_ALL_ITEMS, CONF_UI_CONFIGURATION, DOMAIN
+from .director_utils import update_variables_for_config_entry
+
+_LOGGER = logging.getLogger(__name__)
+
+CONTROL4_POWER_STATE = "POWER_STATE"
+CONTROL4_VOLUME_STATE = "CURRENT_VOLUME"
+CONTROL4_MUTED_STATE = "IS_MUTED"
+CONTROL4_CURRENT_VIDEO_DEVICE = "CURRENT_VIDEO_DEVICE"
+CONTROL4_PLAYING = "PLAYING"
+CONTROL4_PAUSED = "PAUSED"
+CONTROL4_STOPPED = "STOPPED"
+CONTROL4_MEDIA_INFO = "CURRENT MEDIA INFO"
+
+CONTROL4_PARENT_ID = "parentId"
+
+VARIABLES_OF_INTEREST = {
+    CONTROL4_POWER_STATE,
+    CONTROL4_VOLUME_STATE,
+    CONTROL4_MUTED_STATE,
+    CONTROL4_CURRENT_VIDEO_DEVICE,
+    CONTROL4_MEDIA_INFO,
+    CONTROL4_PLAYING,
+    CONTROL4_PAUSED,
+    CONTROL4_STOPPED,
+}
+
+
+class _SourceType(enum.Enum):
+    AUDIO = 1
+    VIDEO = 2
+
+
+@attr.s
+class RoomSource:
+    """Room Source Data."""
+
+    source_type: set[_SourceType] = attr.ib()
+    id: int = attr.ib()
+    name: str = attr.ib()
+
+
+async def get_rooms(hass: HomeAssistant, entry: ConfigEntry):
+    """Return a list of all Control4 rooms."""
+    director_all_items = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR_ALL_ITEMS]
+    return [
+        item
+        for item in director_all_items
+        if "typeName" in item and item["typeName"] == "room"
+    ]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Control4 rooms from a config entry."""
+    all_rooms = await get_rooms(hass, entry)
+    if not all_rooms:
+        return
+
+    entry_data = hass.data[DOMAIN][entry.entry_id]
+    scan_interval = entry_data[CONF_SCAN_INTERVAL]
+    _LOGGER.debug(
+        "Scan interval = %s",
+        scan_interval,
+    )
+
+    async def async_update_data():
+        """Fetch data from Control4 director."""
+        try:
+            return await update_variables_for_config_entry(
+                hass, entry, VARIABLES_OF_INTEREST
+            )
+        except C4Exception as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="room",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=scan_interval),
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
```

Comment:
What happens if this raises `UpdateFailed` or some errors, will the media players still work after a successful refresh?
The same potential issue exists in the light platform.

If this is a problem, maybe the coordinators (this one and the ones in light.py) should all be created in `control4/__init__.py::async_setup_entry` and call `async_config_entry_first_refresh` on them?

Reply:
On reload the integration will work if the director is down during initialization and is brought up before next reload.

Reply:
I don't follow. If we fail to get an update here then the module has failed to setup as we are not able to get a request from controller and would require a reload like any other component. The coordinator is installed for each entity as the poll update mechanism. Failures there would be ignored. Only on setup is this an issue.

Reply:
This change can also be left for another pull request. I have one more change after this to address connection issues that is short and i am hoping we can merge this soon and discuss that in the shorter pull